client: added the "user" option to tasks

mssola opened the pull request:
The `runAs` method already makes sure that the command to be executed is
performed with the given user. Therefore, in this commit this option is just
exported to the configuration of both drivers.

Fixes #276

Signed-off-by: Miquel Sabaté Solà <msabate@suse.com>

mssola commented:
I didn't know what was to best approach to test this in unit tests. As in, maybe I should just use the `nobody` user for `raw_exec` ? But then, what about the `exec` driver which already defaults to "nobody" ?

dadgar commented:
Hey I think we can actually push this up to the Task level as it really applies to all drivers. Let me know if that makes sense or you need any clarification on how to accomplish this

mssola commented:
@dadgar I've updated the patch so this now applies to all drivers. I've tried to keep old behaviors by default. Note though that this PR is not ready to be merged though:

- For whatever reason the job that the agent receives has an empty `User` :/ (I'm new to Nomad's code, so I'm missing something)
- Documentation
- Proper tests (even though I'm still puzzled about what would be the best way to go)

mssola commented:
Ok, I've fixed the issue I was having with the agent and I've also updated the documentation. Any recommendations on providing tests for this ?

I'll squash my commits once everything is ready to be merged :wink:

diptanu commented on client/driver/docker.go:
```diff
@@ -204,6 +203,7 @@ func (d *DockerDriver) containerBinds(alloc *allocdir.AllocDir, task *structs.Ta
 // createContainer initializes a struct needed to call docker.client.CreateContainer()
 func (d *DockerDriver) createContainer(ctx *ExecContext, task *structs.Task,
 	driverConfig *DockerDriverConfig, syslogAddr string) (docker.CreateContainerOptions, error) {
+
```
Please remove the blank line.

diptanu commented on client/driver/utils.go:
```diff
@@ -146,3 +148,12 @@ func GetAbsolutePath(bin string) (string, error) {
+
+// GetExecutorUser returns the user of the task, defaulting to
+// cstructs.DefaultUnprivilegedUser if none was given.
+func GetExecutorUser(task *structs.Task) string {
```
This method probably doesn't need to be exported.

diptanu commented:
@mssola This looks good! Just left a couple of comments. Also, I think we should extend this to the `raw_exec` driver as well.

Regarding tests, you could modify the exec driver and add the User attributes to a few tests and make sure the tests work, the exec driver tests are a little flaky on our local machine currently ( I am investigating). Can you also add a test to make sure the jobspec parsing works?

dadgar commented on website/source/docs/jobspec/index.html.md:
```diff
@@ -228,6 +228,10 @@ The `task` object supports the following keys:
 
+* `user` - Set the user that will run the task. If the driver being used is
+  either `java` or `exec`, then the `nobody` user will be used if this key has
+  not been specified. Otherwise, it defaults to `root`.
```
defaults to the same user the Nomad client is being run as.

dadgar commented on nomad/structs/structs.go:
```diff
@@ -1583,6 +1583,10 @@ type Task struct {
 	// Driver is used to control which driver is used
 	Driver string
 
+	// User is used to determine which user will run the task. It defaults to
```
Same comment

mssola commented:
> Also, I think we should extend this to the raw_exec driver as well.

It's already applying to the `raw_exec` driver. Am I missing something ? :)

@dadgar @diptanu I've added a commit fixing the issues you pointed. Now I'm working on providing tests for this :wink:

mssola commented:
@diptanu I couldn't figure a way around exec that did not look "hackish". So instead it occurred to me to add a test on the docker driver trying to spin up a container with a non-existing user. What do you think about this ?

dadgar commented:
This is awesome! Just one last request, would you mind duplicating the docker test to the other drivers. That way we can ensure we don't regress on any of the drivers!

mssola commented:
@dadgar done. Some tests are failing, but I'm not sure whether this is because my changes :/

dadgar commented:
Wonderful! Thank you so much!
